vision/use-cases/hey_llm/src/main.ts

```diff
@@ -0,0 +1,506 @@
+ * @returns Authorization URL
+ */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function getAuthorizationUrl() {
+  const clientID = PropertiesService.getDocumentProperties().getProperty(
+    PROP_KEY.OAUTH_CLIENT_ID,
+  );
+  const clientSecret = PropertiesService.getDocumentProperties().getProperty(
+    PROP_KEY.OAUTH_CLIENT_SECRET,
+  );
+  if (clientID && clientSecret) {
+    return getGoogleService_().getAuthorizationUrl();
+  }
+  return undefined;
+}
+
+/**
+ * Callback function called after authorization.
+ * @param request Request object
+ * @returns HTML output for the new tab
+ */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function authCallback(request: object) {
+  const isAuthorized = getGoogleService_().handleCallback(request);
+  if (isAuthorized) {
+    return HtmlService.createHtmlOutput('Success! You can close this tab.');
+  } else {
+    return HtmlService.createHtmlOutput('Denied. You can close this tab');
+  }
+}
+/**
+ * Sets OAuth Client ID to Document Properties.
+ * @param val OAuth Client ID
+ */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function setClientID(val: string) {
+  PropertiesService.getDocumentProperties().setProperty(
+    PROP_KEY.OAUTH_CLIENT_ID,
+    val,
+  );
+}
+
+/**
+ * Sets OAuth Client Secret to Document Properties.
+ * @param val OAuth Client Secret
+ */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function setClientSecret(val: string) {
+  PropertiesService.getDocumentProperties().setProperty(
+    PROP_KEY.OAUTH_CLIENT_SECRET,
+    val,
+  );
+}
+
+/**
+ * Asks Gemini for a response based on a provided context and input.
+ * @param {string} instruction An instruction that describes the task.
+ * @param {string} input The input data to process based on the instruction.
+ * @param {string[][]} context The context that LLM should be aware of.
+ * @return {string} The LLM's response, trimmed and ready for use in a spreadsheet cell.
+ * @customFunction
+ */
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function HEY_LLM(instruction: string, input: string, context: string[][] = []) {
+  const prompt = `
+## Instruction
+${instruction}
+
+## Context (CSV formatted)
+${context ? context.map(row => row.join(', ')).join('\n') : ''}
+
+## Task
+Input: ${input}
+Output:`;
+
+  const cache = CacheService.getDocumentCache();
+  const cacheKey = `hey_llm:${generateHashValue(prompt)}`;
+  const cached = cache?.get(cacheKey);
+  if (cached) return cached;
+
+  const oauthService = getGoogleService_();
+  const url = `https://${LOCATION}-aiplatform.googleapis.com/v1/projects/${getProjectNumber_()}/locations/${LOCATION}/publishers/google/models/${GEMINI_MODEL}:generateContent`;
+
+  const payload = JSON.stringify({
+    contents: [
+      {
+        role: 'user',
+        parts: [
+          {
+            text: prompt,
+          },
+        ],
+      },
```

Comment:
Is it possible to use the Node.JS SDK https://cloud.google.com/vertex-ai/generative-ai/docs/reference/nodejs/latest

Instead of constructing the request manually?

Reply:
I need to construct the request manually because AppScript doesn't have the global `fetch` method and all external network requests should be made via `UrlFetchApp` service.